sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -140,14 +141,18 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper wit
                         val count = aggregate.Count(avg.child).toAggregateExpression(isDistinct)
                         // Closely follow `Average.evaluateExpression`
                         avg.dataType match {
+                          case dt: DecimalType if avg.failOnError =>
+                            addCastIfNeeded(DecimalPrecision.decimalAndDecimal()(
+                              Divide(
+                                CheckOverflowInSum(sum, dt, false),
+                                addCastIfNeeded(count, dt), failOnError = false)), avg.dataType)
                           case _: YearMonthIntervalType =>
                             If(EqualTo(count, Literal(0L)),
                               Literal(null, YearMonthIntervalType()), DivideYMInterval(sum, count))
                           case _: DayTimeIntervalType =>
                             If(EqualTo(count, Literal(0L)),
                               Literal(null, DayTimeIntervalType()), DivideDTInterval(sum, count))
                           case _ =>
```

Comment:
I think we need to check ansi mode in this case as well, as avg(long) can also overflow